Feature/models

noadudai opened the pull request.

noadudai committed:
Added a few models

Added an EmployeeModel, DayModel, ShiftsModel models and 2 enums to add constants, making the code more readable and less error-prone.

noadudai committed:
preferences list in employee model

noadudai committed:
week and constraint model

noadudai committed:
fixed circular import

timopomer commented on Models/Constraint_model.py:
```diff
@@ -0,0 +1,24 @@
+                for shift in self.week_info.days_in_this_week[day].shifts:
+
+                    day_name = self.week_info.days_in_this_week[day].today_is.value
+                    # print(day_name, employee.name, shift.type)
```
why is this commented out?

timopomer commented on Models/Constraint_model.py:
```diff
@@ -0,0 +1,24 @@
+        self.all_possible_shifts_combo = {}
+        self._populate_all_possible_shifts_combo()
+
+    def _populate_all_possible_shifts_combo(self):
```
what does this return?

timopomer commented on Models/Constraint_model.py:
```diff
@@ -0,0 +1,24 @@
+from ortools.sat.python import cp_model
+
+
+class ConstraintsModel:
```
why is this a class? does it need to keep a state?

timopomer commented on Models/Day_model.py:
```diff
@@ -0,0 +1,18 @@
+
+class DayModel:
+
+    def __init__(self, day_of_the_week, shifts):
```
use dataclasses for stuff like this. gives you also comparisons out of the box

timopomer commented on Models/Employee_model.py:
```diff
@@ -0,0 +1,27 @@
+from enum import Enum
+from typing import List
+#  Priority levels
+class PriorityEnum(Enum):
```
module per class

timopomer commented on Models/Employee_model.py:
```diff
@@ -0,0 +1,27 @@
+    def update_priority(self, new_priority):
+        self.priority = new_priority
+
+    def update_status(self, status: bool):
```
status being a boolean? thats pretty confusing. 

`employee.update_status(True)` doesnt syntaxically mean anything

timopomer commented on Models/Employee_model.py:
```diff
@@ -0,0 +1,27 @@
+    def update_status(self, status: bool):
+        self.is_new = status
+
+    def update_preferences(self, new_preferences):
```
why are there setters like this? seems to me like this class can also be a dataclass.

timopomer commented on Models/Shifts_model.py:
```diff
@@ -0,0 +1,9 @@
+class ShiftsModel:
+
+    def __init__(self, shift_type: str):
+        self.type = shift_type
```
why is the `shift_type` a string?

timopomer commented on Models/Shifts_model.py:
```diff
@@ -0,0 +1,9 @@
+class ShiftsModel:
```
why are all the classes called "model" ?

timopomer commented on Models/Employee_model.py:
```diff
@@ -0,0 +1,27 @@
+
+class EmployeeModel:
+
+    def __init__(self, name: str, priority, is_new: bool, preferences):
```
what the hell is `priority` and `preferences`

timopomer commented on Models/Day_model.py:
```diff
@@ -0,0 +1,18 @@
+from enum import Enum
+
+
+class DaysEnum(Enum):
```
one class per module. the convention is modules are `snake_case`, classes `PascalCase`

noadudai committed:
Fixed CR questions

noadudai committed:
Fixed CR questions

timopomer commented on models/constraints.py:
```diff
@@ -0,0 +1,42 @@
+
+class Constraints:
+
+    def __init__(self, week_info, employees):
```
types

timopomer commented on models/constraints.py:
```diff
@@ -0,0 +1,42 @@
+
+                for shift in week_info.days_in_this_week[day].shifts:
+
+                    all_possible_shifts_combo[(employee.name, day_name, shift.type)] = model.NewBoolVar(
```
two employees with the same name?

timopomer commented on models/constraints.py:
```diff
@@ -0,0 +1,42 @@
+        return all_possible_shifts_combo
+
+    # A constraint that ensures that there will be only one employee in each shift per day
+    def one_employee_in_each_shift_constraint(self):
```
functions are verbs

timopomer commented on models/days/day.py:
```diff
@@ -0,0 +1,12 @@
+class Day:
+
+    today_is: DaysEnum
+    shifts: List[Shift]
```
isnt this problematic? shift can be across days

timopomer commented on models/days/day.py:
```diff
@@ -0,0 +1,12 @@
+@dataclass
+class Day:
+
+    today_is: DaysEnum
```
`current_day` ?
`weekday`?

timopomer commented on models/employees/employee_status_enum.py:
```diff
@@ -0,0 +1,6 @@
+
+
+class EmployeeStatusEnum(Enum):
+    new_employee = True
```
what? why booleans? is there nothing in between?

timopomer commented on models/shifts/shift.py:
```diff
@@ -0,0 +1,12 @@
+
+    def __init__(self, shift_type: ShiftsEnum):
+        self.type = shift_type
+        self.employee = None
```
why arent shifts generated once its clear what they will contain? dont like classes that change too much after being created

timopomer commented on models/shifts/shifts_enum.py:
```diff
@@ -0,0 +1,10 @@
+from enum import Enum
+
+
+class ShiftsEnum(Enum):
```
`ShiftTypesEnum`

timopomer commented on models/week.py:
```diff
@@ -0,0 +1,9 @@
+
+@dataclass
+class Week:
+    days_in_this_week: List[Day]
```
`weekdays` ?

timopomer commented on .idea/inspectionProfiles/profiles_settings.xml:
```diff
@@ -1,5 +1,6 @@
 <component name="InspectionProjectProfileManager">
   <settings>
+    <option name="PROJECT_PROFILE" value="Default" />
```
gitignore

timopomer replied:
you need to remove this file from the commit. find out how to do it

noadudai committed:
enum and not str

timopomer commented on models/constraints.py:
```diff
@@ -0,0 +1,42 @@
+    def _generate_all_possible_shifts_combo(employees, week_info, model):
+        all_possible_shifts_combo = {}
+        for employee in employees:
+            for day in range(len(week_info.days_in_this_week)):
```
something about this screams `C` to me. why not just iterate the weekdays?

noadudai committed:
Added 2 constraints.

Removed Day, Week, and DaysEnum classes, working on only shift and employee classes for better data management

noadudai committed:
ShiftTypesEnum

noadudai committed:
add_no_morning_shift_after_closing_shift_constraint

noadudai committed:
employee status and refactoring main test

noadudai committed:
add_no_more_that_6_working_days_a_week_constraint

noadudai committed:
todos

timopomer commented on main.py:
```diff
@@ -0,0 +1,66 @@
+from models.constraints import Constraints
+
+
+def sun_to_sat(sun: int, sat: int, month: int, year: int, shifts):
```
function not a verb

timopomer commented on main.py:
```diff
@@ -0,0 +1,66 @@
+                shifts.append(Shift(ShiftTypesEnum.CLOSING, datetime(year, month, day, 19, 30), datetime(year, month, day + 1, 2)))
+
+
+def test():
```
write tests. or call this main

timopomer commented on models/constraints.py:
```diff
@@ -0,0 +1,172 @@
+        all_possible_shift_employee_combinations = {}
+        for employee in employees:
+            for shift in shifts:
+                all_possible_shift_employee_combinations[
```
making this a oneline makes you
1. repeat yourself like in `str(shift.start_date_and_time_of_shift.date())`
2. unreadable.

create variables with useful names and append to the model in a small line

timopomer commented on models/constraints.py:
```diff
@@ -0,0 +1,172 @@
+
+    # A constraint that ensures that there will be only one employee in each shift per day
+
+    def add_only_one_employee_in_each_shift_constraint(self):
```
why is this one not static and accessing `self.employees` and `_generate_all_possible_shift_employee_combinations` isnt?

timopomer replied:
for other funcs here too

timopomer commented on models/constraints.py:
```diff
@@ -0,0 +1,172 @@
+                shifts_on_day = []
+
+                for shift in self.shifts:
+                    if str(shift.start_date_and_time_of_shift.date()) == day:
```
`str(shift.start_date_and_time_of_shift.date())` DRY principle

timopomer replied:
im surprised `day` is a string

timopomer commented on models/constraints.py:
```diff
@@ -0,0 +1,172 @@
+
+            for employee in self.employees:
+                for shift in self.shifts:
+                    if str(shift.start_date_and_time_of_shift.date()) == day and shift.type == first_shift and employee.status == EmployeeStatusEnum.new_employee:
```
these two lines are very similar. share more code. you can write nested functions for this with useful names

timopomer commented on models/constraints.py:
```diff
@@ -0,0 +1,172 @@
+            worked_closing_shift_yesterday = self.cp_model.NewBoolVar(f"{employee.name}_worked_closing_shift_yesterday_day")
+
+            # checking for every shift if the shift before was a closing shift
+            for shift in range(1, len(self.shifts)):
```
you can iterate the shifts in pairs. look into itertools

noadudai committed:
CR pints fixed

timopomer commented on constraints_file.py:
```diff
@@ -0,0 +1,189 @@
+
+
+# A function to return a string representing the start date of the given shift
+def get_str_start_date_from_shift(shift):
```
shouldnt  this be in `Shift` ?

timopomer replied:
why dont you use `get_start_date_of_shift` here

timopomer commented on constraints_file.py:
```diff
@@ -0,0 +1,189 @@
+                )
+
+
+def objective_function(shifts: list[Shift], employees: List[Employee], constraint_model, shift_combinations):
```
this function is like a test function, it doesnt return anything useful

timopomer commented on main.py:
```diff
@@ -0,0 +1,66 @@
+    objective_function(shift_this_week, employees, cp_model, all_shifts)
+
+
+    print("Hello World!")
```
garbage

timopomer commented on models/shifts/shift.py:
```diff
@@ -0,0 +1,15 @@
+        self.end_date_and_time_of_shift = end_date_and_time_of_shift
+
+    # Setting the employee that will work this shift after the schedule has been created
+    def set_employee(self, employee):
```
this is never called by anybody.

noadudai committed:
rremoved files that should not be in git

timopomer commented on setup.py:
```diff
@@ -7,7 +7,8 @@
       url='https://github.com/noadudai/EmployeeShiftsSchedule',
       install_requires=[
-            'ortools==9.7.2996',
+            'ortools',
```
any version?

timopomer commented on .idea/inspectionProfiles/profiles_settings.xml:
fix this

timopomer commented on constraints_file.py:
```diff
@@ -0,0 +1,189 @@
+#       (as a datetime object) of the shift, end date ( also as a datetime object) of the shift, and the shift's type,
+#       as a key, and the value will be a boolean that will represent true or false based of if the employee is
+#       working that shift or not.
+def generate_shift_employee_combinations(employees: List[Employee], shifts: list[Shift], constraint_model):
```
its a good idea to create return types

timopomer commented on constraints_file.py:
```diff
@@ -0,0 +1,189 @@
+
+
+# A constraint that ensures that each employee does not work more than 6 days in a week
+def add_max_6_working_days_a_week_constraint(shifts: list[Shift], employees: List[Employee], constraint_model, shift_combinations):
```
make the max working days a parameter?

timopomer commented on constraints_file.py:
```diff
@@ -0,0 +1,189 @@
+    pass
+
+
+# A constraint that ensures that an employee how is working on a closing shift,
```
ingles

timopomer commented on main.py:
```diff
@@ -0,0 +1,66 @@
+
+    # datetime.weekday() will return an int representing: Monday is 0 and Sunday is 6
+
+    for day in range(sun, sat + 1):
```
this is weird to me, why are  `sun`, `sat` input parameters?

timopomer commented on main.py:
```diff
@@ -0,0 +1,66 @@
+
+    for day in range(sun, sat + 1):
+        # if the day is fri or sat
+        if 3 < datetime(year, month, day).weekday() < 6:
```
consider `SUN, MON, TUE, WED, THU, FRI, SAT = range(7)`
or creating a readable enum

timopomer commented on main.py:
```diff
@@ -0,0 +1,66 @@
+            shifts.append(Shift(ShiftTypesEnum.CLOSING, datetime(year, month, day, 21, 30), datetime(year, month, day + 1, 2)))
+
+        # sun to thu
+        else:
```
why are these hardcoded?

noadudai committed:
Added WorkersSchedule class for the resulting schedule

noadudai committed:
cr fix

removed hardcoded times, need to do a better solution.

timopomer commented on models/shifts/shifts_times.py:
```diff
@@ -0,0 +1,16 @@
+
+# Need to think of a better solution for the shifts start time and end time storing or receiving form.
+class ShiftTimes(Enum):
+    morning_shift_times = {"start": time(11, 45), "end": time(17, 0)}
```
not an enum.

timopomer commented on models/workers_schedule.py:
```diff
@@ -0,0 +1,10 @@
+
+class WorkersWeekSchedule:
+    
+    def __init__(self, week_schedule: list):
```
use `List` then

timopomer commented on models/shifts/shift.py:
```diff
@@ -0,0 +1,26 @@
+class Shift:
+
+    def __init__(self, shift_type: ShiftTypesEnum, start_date_and_time_of_shift: datetime, end_date_and_time_of_shift: datetime):
+        self.type = shift_type
```
dangerous to override `self.type`

timopomer commented on constraints_file.py:
```diff
@@ -0,0 +1,170 @@
+    for employee in employees:
+        worked_closing_shift_yesterday = constraint_model.NewBoolVar(f"{employee.id}_worked_closing_shift_yesterday_day")
+
+        for shift_pair in pairwise(shifts):
```
usually we keep the `itertools.pairwise`

timopomer commented on main.py:
```diff
@@ -0,0 +1,95 @@
+
+
+# A function to generate the shifts from sunday to saturday
+def generate_sun_to_sat_shifts(sun_day: int, sat_day: int, month: int, year: int, shifts):
```
just get a `timedelta` and `datetime`

add a type to `shifts`

noadudai committed:
cr fix

Added a shift types fie for the shift types with type start time and end time values

timopomer commented on main.py:
```diff
@@ -0,0 +1,85 @@
+
+    delta = datetime.timedelta(days=1)
+
+    while day <= saturday:
```
list comprehension

timopomer commented on constraints_file.py:
```diff
@@ -0,0 +1,167 @@
+
+
+# A constraint that ensures that each employee does not work more than 6 days in a week
+def add_max_working_days_a_week_constraint(shifts: list[Shift], employees: List[Employee], constraint_model, shift_combinations, max_working_days) -> None:
```
broker

timopomer commented on main.py:
```diff
@@ -0,0 +1,85 @@
+        day += delta
+
+
+if __name__ == "__main__":
```
not callable

noadudai committed:
Tests

timopomer commented on test_constraints.py:
put in tests folder

timopomer commented on test_constraints.py:
```diff
@@ -0,0 +1,383 @@
+from models.shifts.shifts_types_enum import ShiftTypesEnum
+
+
+def test_add_one_employee_per_shift_constraint_broken_down():
```
broken down?

timopomer commented on test_constraints.py:
```diff
@@ -0,0 +1,383 @@
+                print(employee.id, employee.status.value, shift.shift_type.name_of_shift.value, shift.get_str_start_date())
+
+
+def test_all_constraints2():
```
weird function name

timopomer commented on test_constraints.py:
```diff
@@ -0,0 +1,383 @@
+    schedule = WorkersWeekSchedule(schedule_shifts)
+
+    # printing to see that the solution is satisfying all the constraints
+    print("")
```
print in test? what does this assert?

timopomer commented on test_constraints.py:
```diff
@@ -0,0 +1,383 @@
+    for shift in shift_this_week:
+        for employee in employees:
+            if solver.Value(shifts[(employee.id, shift.get_str_start_date(), shift.get_str_end_date(), shift.shift_type.name_of_shift.value)]):
+                print(employee.id, employee.status.value, shift.shift_type.name_of_shift.value, shift.get_str_start_date())
```
this doesnt test anything

timopomer commented on constraints.py:
```diff
@@ -0,0 +1,157 @@
+
+
+# Returns a list containing all the "new employees" of each shift
+def generate_new_employees_on_shift(employees, shifts, shift_combinations, shift_type: ShiftTypesEnum, day) -> list:
```
`day` type?

timopomer replied:
returns list of what?

timopomer commented on constraints.py:
```diff
@@ -0,0 +1,157 @@
+
+
+# A constraint that ensures that there will be only one employee in each shift per day
+def add_one_employee_per_shift_constraint(shifts: list[Shift], employees: List[Employee], constraint_model: cp_model.CpModel, shift_combinations) -> None:
```
type of `shift_combinations` ?

timopomer commented on constraints.py:
```diff
@@ -0,0 +1,157 @@
+                start_closing_shift = first_shift.get_str_start_date()
+                end_closing_shift = first_shift.get_str_end_date()
+
+                start_morning_shift = second_shift.get_str_start_date()
```
why is this a morning shift?

noadudai committed:
Fixed tests

timopomer commented on test/test_constraints.py:
```diff
@@ -0,0 +1,339 @@
+    assert (max_working_days_in_a_week >= max(counter_dict_schedule.values()))
+
+    for pair in itertools.pairwise(schedule.week_schedule):
+        first_shift, second_shift = pair
```
deconstruct

timopomer commented on test/test_constraints.py:
```diff
@@ -0,0 +1,339 @@
+                schedule_shifts.append([shift, employee.id])
+    schedule = WorkersWeekSchedule(schedule_shifts)
+
+    # printing to see that the solution is satisfying all the constraints
```
printing?

timopomer commented on test/test_constraints.py:
```diff
@@ -0,0 +1,339 @@
+
+    # printing to see that the solution is satisfying all the constraints
+    dict_schedule = dict(schedule.week_schedule)
+    counter_dict_schedule = Counter(dict_schedule.values())
```
Counter? whats that

timopomer commented on test/test_constraints.py:
```diff
@@ -0,0 +1,339 @@
+    # there is a solution where employees are not assigned to shifts, which is considered an optimal solution.
+    assert (status == cp_model.OPTIMAL)
+
+    if status == cp_model.OPTIMAL:
```
what?

timopomer commented on test/test_constraints.py:
```diff
@@ -0,0 +1,339 @@
+    assert (status == cp_model.OPTIMAL)
+
+
+def test_add_one_employee_per_shift_constraint_with_no_employees():
```
what?

timopomer commented on constraints.py:
```diff
@@ -0,0 +1,158 @@
+
+# Returns a list of boolvars representing a true false of all the "new employees" that work each shift
+# the day parameter is the str date of the given day.
+def generate_new_employees_on_shift(employees: list[Employee], shifts: list[Shift], shift_combinations: dict[tuple, bool], shift_type: ShiftTypesEnum, day: str) -> list:
```
why would `day` be a `str`

timopomer replied:
`list` of what?

timopomer commented on constraints.py:
```diff
@@ -0,0 +1,158 @@
+#       (as a datetime object) of the shift, end date ( also as a datetime object) of the shift, and the shift's type,
+#       as a key, and the value will be a boolean that will represent true or false based of if the employee is
+#       working that shift or not.
+def generate_shift_employee_combinations(employees: List[Employee], shifts: list[Shift], constraint_model: cp_model.CpModel) -> dict:
```
`dict` of what

timopomer commented on constraints.py:
```diff
@@ -0,0 +1,158 @@
+
+
+# A constraint that ensures that each employee works at most one shift per day
+def add_at_most_one_shift_in_the_same_day_constraint(shifts: list[Shift], employees: List[Employee], constraint_model: cp_model.CpModel, shift_combinations) -> None:
```
what is `shift_combinations`

timopomer commented on constraints.py:
```diff
@@ -0,0 +1,158 @@
+    # simple boolean expression, the solver knows which BoolVar variables are part of the sum expression and enforces
+    # the constraint based on the values of those variables.
+    for employee in employees:
+        constraint_model.Add(sum(shift_combinations[(employee.id, shift.get_str_start_date(), shift.get_str_end_date(), shift.shift_type.name_of_shift.value)] for shift in shifts) <= max_working_days)
```
sub vars

timopomer commented on test/test_constraints.py:
```diff
@@ -0,0 +1,339 @@
+
+    # the second employee does not work in the same shift as the first employee, it chose the first employee to
+    # work that shift because he is the first value in the given employee list
+    assert (solver.Value(shifts[(
```
variables

noadudai committed:
sub variables

timopomer commented on test/test_constraints.py:
```diff
@@ -0,0 +1,348 @@
+        if shift_in_first_pair.shift_type.name_of_shift == ShiftTypesEnum.EVENING and shift_in_second_pair.shift_type.name_of_shift == ShiftTypesEnum.CLOSING:
+            first_employee = None
+            second_employee = None
+            for employee in employees:
```
why

timopomer commented on test/test_constraints.py:
```diff
@@ -0,0 +1,348 @@
+    assert (status == cp_model.OPTIMAL)
+
+    second_employee_assignment = shifts[(test_employee2.id, start_shift, end_shift, shift_type)]
+    assert (solver.Value(second_employee_assignment) == 0)
```
why 0

noadudai committed:
cr fix
